This script contains the following points: 
1. import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.
2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.
3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?
4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.
5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.
6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying.
7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.
8. Export the newly created dataframe as ords_prods_merge in a suitable format.

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

 1.import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [2]:
#import .csv
path=r'E:\Instacart Basket Analysis'
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_updated.pkl'))

2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [4]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.393473
2                17.000546
3                17.216578
4                17.714601
5                14.958409
6                16.454727
7                17.325861
8                15.117326
9                16.027829
10               20.229811
11               16.453444
12               15.895358
13               16.692698
14               16.817133
15               16.133323
16               17.607401
17               15.936598
18               19.646685
19               17.216008
20               16.393918
21               22.580576

3.Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

When comparing the results obtained from a subset of data with the results obtained from the entire dataset, we notice that the latter includes all 'department_id' values. Moreover, the mean values of the department ids present in the subset have changed, with some increasing and others decreasing. For instance, in the subset, department id 20 had a mean value of '17.251314', while in the entire dataframe, it has a mean value of '16.393918'.b

4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

a)Split the data into groups based on the “user_id” column.
b)Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
c)Create a new column, “max_order,” into which you’ll place the results of your aggregation.
d)creat loyalty flag with loc().

In [5]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform("max")

In [6]:
ords_prods_merge.head(100)

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
1            1   2398795        1    prior             2                   3   
2            2    473747        1    prior             3                   3   
3            3   2254736        1    prior             4                   4   
4            4    431534        1    prior             5                   4   
..         ...       ...      ...      ...           ...                 ...   
95          95   3226575      360    prior             1                   5   
96          96   1469869      377    prior             3                   5   
97          97   1927023      387    prior             2                   4   
98          98    858092      420    prior             4                   1   
99          99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8               11.114836         196                  1   
1                   7               15.000000         196                  1   
2                  12               21.000000         196                  1   
3                   7               29.000000         196                  1   
4                  15               28.000000         196                  1   
..                ...                     ...         ...                ...   
95                 12               11.114836         196                  1   
96                 17                3.000000         196                  9   
97                 10               22.000000         196                  3   
98                 19               30.000000         196                  2   
99                 19               16.000000         196                  2   

    ...  product_name aisle_id department_id  prices  merge_flag  \
0   ...          Soda       77             7     9.0        both   
1   ...          Soda       77             7     9.0        both   
2   ...          Soda       77             7     9.0        both   
3   ...          Soda       77             7     9.0        both   
4   ...          Soda       77             7     9.0        both   
..  ...           ...      ...           ...     ...         ...   
95  ...          Soda       77             7     9.0        both   
96  ...          Soda       77             7     9.0        both   
97  ...          Soda       77             7     9.0        both   
98  ...          Soda       77             7     9.0        both   
99  ...          Soda       77             7     9.0        both   

       busiest day     busiest_day    busiest_days busiest_period_of_day  \
0   Regularly busy  Regularly busy  Regularly busy           Most orders   
1   Regularly busy  Regularly busy  Regularly busy         Fewest orders   
2   Regularly busy  Regularly busy  Regularly busy           Most orders   
3       Least busy      Least busy      Least busy         Fewest orders   
4       Least busy      Least busy      Least busy           Most orders   
..             ...             ...             ...                   ...   
95  Regularly busy  Regularly busy  Regularly busy           Most orders   
96  Regularly busy  Regularly busy  Regularly busy         Fewest orders   
97      Least busy      Least busy      Least busy           Most orders   
98  Regularly busy  Regularly busy  Regularly busy         Fewest orders   
99  Regularly busy  Regularly busy  Regularly busy         Fewest orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10  
..       ...  
95         3  
96         3  
97         8  
98        22  
99        22  

[100 rows x 22 columns]

In [7]:
pd.options.display.max_rows = None

In [8]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    4865587
Loyal customer      3154279
New customer        1898387
Name: count, dtype: int64

5.The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [12]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'std', 'median']})

prices                                 
                       mean  min      max         std median
loyalty_flag                                                
Loyal customer     9.435027  1.0  99999.0  227.642420    7.4
New customer      12.705686  1.0  99999.0  553.163202    7.4
Regular customer  11.675800  1.0  99999.0  448.486239    7.4

6.The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
a)If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
b)If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [13]:
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [14]:
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

In [15]:
ords_prods_merge.loc[ords_prods_merge['average_spend']>=10,'spender_flag']='High_spender'

In [16]:
ords_prods_merge['spender_flag'].value_counts(dropna=False)

spender_flag
Low_spender     9736335
High_spender     181918
Name: count, dtype: int64

7.In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
a)If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
b)If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
c)If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [17]:
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [18]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency']>20,'frequency_flag']='Non-frequent customer.'

In [19]:
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20),'frequency_flag']='Regular customer.'

In [20]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency']<=10,'frequency_flag']='Frequent customer.'

In [22]:
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer.        6258107
Regular customer.         2840401
Non-frequent customer.     819745
Name: count, dtype: int64

8. Export the newly created dataframe as ords_prods_merge in a suitable format.

In [23]:
# Export data to pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_grouping_customers.pkl'))